openstudiocore/src/energyplus/ForwardTranslator/ForwardTranslateElectricEquipmentITEAirCooled.cpp

```diff
@@ -0,0 +1,265 @@
+*  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+*  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+*  STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+*  ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+***********************************************************************************************************************/
+
+#include "../ForwardTranslator.hpp"
+
+#include "../../model/Model.hpp"
+#include "../../model/ElectricEquipmentITEAirCooled.hpp"
+#include "../../model/ElectricEquipmentITEAirCooled_Impl.hpp"
+#include "../../model/ElectricEquipmentITEAirCooledDefinition.hpp"
+#include "../../model/ElectricEquipmentITEAirCooledDefinition_Impl.hpp"
+#include "../../model/Space.hpp"
+#include "../../model/Space_Impl.hpp"
+#include "../../model/SpaceType.hpp"
+#include "../../model/SpaceType_Impl.hpp"
+#include "../../model/ThermalZone.hpp"
+#include "../../model/ThermalZone_Impl.hpp"
+#include "../../model/Schedule.hpp"
+#include "../../model/Schedule_Impl.hpp"
+#include "../../model/Curve.hpp"
+#include "../../model/Curve_Impl.hpp"
+#include "../../model/CurveBiquadratic.hpp"
+#include "../../model/CurveBiquadratic_Impl.hpp"
+#include "../../model/CurveQuadratic.hpp"
+#include "../../model/CurveQuadratic_Impl.hpp"
+#include "../../model/Node.hpp"
+#include "../../model/Node_Impl.hpp"
+#include "../../model/PortList.hpp"
+#include "../../model/PortList_Impl.hpp"
+#include "../../model/AirTerminalSingleDuctVAVNoReheat.hpp"
+#include "../../model/AirTerminalSingleDuctVAVNoReheat_Impl.hpp"
+#include "../../model/AirTerminalSingleDuctVAVReheat.hpp"
+#include "../../model/AirTerminalSingleDuctVAVReheat_Impl.hpp"
+#include "../../model/HVACComponent.hpp"
+#include "../../model/LifeCycleCost.hpp"
+
+#include <utilities/idd/ElectricEquipment_ITE_AirCooled_FieldEnums.hxx>
+#include "../../utilities/idd/IddEnums.hpp"
+#include <utilities/idd/IddEnums.hxx>
+
+using namespace openstudio::model;
+
+using namespace std;
+
+namespace openstudio {
+
+namespace energyplus {
+
+boost::optional<IdfObject> ForwardTranslator::translateElectricEquipmentITEAirCooled( ElectricEquipmentITEAirCooled & modelObject )
+{
+  IdfObject idfObject(openstudio::IddObjectType::ElectricEquipment_ITE_AirCooled);
+  m_idfObjects.push_back(idfObject);
+
+  for (LifeCycleCost lifeCycleCost : modelObject.lifeCycleCosts()){
+    translateAndMapModelObject(lifeCycleCost);
+  }
+
+  ElectricEquipmentITEAirCooledDefinition definition = modelObject.electricEquipmentITEAirCooledDefinition();
+
+  // Pre-process: 
+  // (1) assign to zone; (2) connect the supply air node (3) apply contraints
+  // ITE object that is assigned to space type in OS has been switched to separate spaces in ForwardTranslator
+  boost::optional<Space> space = modelObject.space();
```

Comment:
I thought about this too. Let alone connecting the node, E+ will crash when no space is assigned to the object. However, I checked ForwardTranslateLights for reference, which didn't assert that either a space or spacetype has to be initialized and associated.
[https://github.com/NREL/OpenStudio/blob/develop/openstudiocore/src/energyplus/ForwardTranslator/ForwardTranslateLights.cpp#L65](url)  So I assume you let this pass to E+ error? Not sure if I misunderstand, please correct me if I'm wrong.

Now about connecting the node, a space is needed for node extraction ONLY if the Calculation Method = "FlowControlWithApproachTemperatures" or if the "Air Inlet Connection Type" is "AdjustedSupply". Otherwise, however, the object can proceed without a space. But eventually E+ will still crash if no space is assigned to the object.

If it is true that you are letting the "no space assignment" pass to E+ error, I guess I should not add assert here.

Reply:
@macumber Any thoughts?